vmnet.framework support on macOS

NattyNarwhal opened the pull request:
macOS has no built-in tun/tap support; third-party extensions to do so require more hoops to jump through as Apple (justifiably) makes it harder to use third-party kexts. The situation for virtual networking is thus clumsy, relying on pcap/VDE, or dealing with the limitations of slirp or the UDP tunneling. This PR adds support for [vmnet](https://developer.apple.com/documentation/vmnet?language=objc), an alternative built into the OS. It provides host-only and shared NAT layer 2 virtual networks, as well as bridging virtual interfaces to physical interfaces.

vmnet requires macOS 10.10 for basic functionality, and 10.15 for bridged networks.

Documentation and SCP help has also been updated.

This has been tested with the following scenarios on macOS 15:

- Netbooting a MicroVAX 3900 in simh from an InfoServer 150 in simh (over MOP); the system was able to install OpenVMS 7.3.
- Installing NetBSD/vax 10.1. NetBSD is able to acquire IPv4 and IPv6 addresses via DHCP, and I can telnet into the VAX via IPv6.

NattyNarwhal committed:
Initial support for vmnet.framework

On macOS, tap devices for L2 networking are not supported out of the
box. While a kext can be added to provide tap support, the kext
experience is not very good; Apple has strongly recommended against
their usage.

As a replacement that's documented and recommended, Apple introduced the
vmnet framework, intended for emulators and virtualization software
explicitly. This API requires macOS 10.10, with bridged network support
coming in macOS 10.15.

This introduces basic support for vmnet.framework in SIMH. I've tested
it by booting an emulated MicroVAX 3800 from an emulated InfoServer 150,
where it was able to reach OpenVMS 7.3 standalone BACKUP.

NattyNarwhal committed:
Allow for shared/bridged mode, show in SHOW ETHERNET

Show the allowed bridged network devices SHOW ETHERNET, as well as the
shared/host networking modes. It also shows shared/host/device name for
bridge in i.e. "HELP XS" output as well.

Shared, bridged, and host modes have other options for configuration;
i.e. isolation, IP ranges, etc. Some of these are not well documented,
so should look into these. Bridged mode needs macOS 10.15.

NattyNarwhal committed:
Check for macOS 10.15 to used bridge mode, make usage clear

Bridged devices require macOS 10.15 to work correctly. Check if we're on
macOS 10.15 at runtime (so a newer SDK can be used to make simh binaries
that target older macOS), and at compile time (so an older SDK can be
used to build simh).

Also make it so vmnet: for host only is deprecated; you must explicitly
provide either host, shared, or a bridged device name.

NattyNarwhal committed:
Defensive checks for vmnet.framework

Require targeting macOS 10.10, in case the user builds with an older SDK
or sets the target version to an older release. This shouldn't be a
problem with modern macOS SDKs, but users using or targeting older macOS
should have a better message available rather than a compile message.
The check is ugly, but it should work.

A similar check is done for block support; this should only be a problem
with users using GCC instead of clang, which is default on macOS.

Also add a message for vmnet support when printing the build config.

NattyNarwhal committed:
Move vmnet below the definition of MAX to avoid redefinition

NattyNarwhal committed:
vmnet.framework with legacy makefiles

I have not put as many checks as the CMake version has. Those may get
unwieldy.

NattyNarwhal committed:
Initial pass at documenting vmnet usage

NattyNarwhal commented:
The ethernet README mentions it, but it's possible to provide additional knobs; vmnet supports them, it's mostly an issue of figuring out how we want to provide the configuration for it.

markpizz commented:
If what you've proposed here is, in fact, functionally equivalent to existing ethernet functionality (i.e. pcap, tun/tap, nat), then the simh user commands used to configure these things should not be any different than the current configuration steps.
If along with what is mentioned in my first sentence additional capabilities are also available (i.e. bridging) then extended commands should be added for this functionality,

NattyNarwhal commented:
> If what you've proposed here is, in fact, functionally equivalent to existing ethernet functionality (i.e. pcap, tun/tap, nat), then the simh user commands used to configure these things should not be any different than the current configuration steps.

Correct, like the other alternate ethernet backends, it's just prefixed (i.e. `vmnet:host` or `vmnet:en0`). It is a little different in that bridging and NAT are presented under the same interface (i.e. it can be used as a substitute for slirp or pcap).

> If along with what is mentioned in my first sentence additional capabilities are also available (i.e. bridging) then extended commands should be added for this functionality,

Right, the vmnet API provides some additional configuration options, but I'm wondering how best to present the options (i.e. a larger connection string, or additional `SET` options?).

----

One thing I also forgot to mention is the MAC address behaviour. When creating a vmnet interface, it gets a MAC address assigned, and this can be overridden (`vmnet_interface_id_key` key on the interface description). It doesn't seem you need to have the emulated system use the vmnet interface's MAC address though. I've been using simh with the assigned MAC addresses for VAXen and it seems fine, so I'm not sure if overriding the automatically assigned vmnet MAC address is useful. It does make `SHOW ETHERNET` mildly more useful (since you can xref vmnet interfaces in i.e. ifconfig w/ it).

For the purposes of vmnet in simh, this does mean changing the MAC address handling as to set the MAC address when the interface is created (rather than look it up later).

NattyNarwhal committed:
Keep track of generated vmnet MAC addr

The vmnet interface created for a simh instance has its own MAC address.
It doesn't seem too useful to for simh itself (you can just use whatever
MAC address and send/receive frames with that), but it does make
referencing it in other places (i.e. SHOW ETHERNET vs. ifconfig) easier.

The generated MAC address is provided at interface creation time; while
we could look it up earlier like with pcap, this seems more efficient.

pkoning2 commented:
Looking at the "vmnet" web documentation you cited I wonder how it applies to SIMH.  It's described as an API for hypervisors to construct virtual interfaces, which have their own IP addresses and other stuff.  How does that fit what SIMH needs, which is an ethernet level send/receive service?
I agree that tun/tap has long been problematic which is why I stopped looking at it years ago.  But pcap works just fine at least on Mac and Linux.

NattyNarwhal commented:
> Looking at the "vmnet" web documentation you cited I wonder how it applies to SIMH. It's described as an API for hypervisors to construct virtual interfaces, which have their own IP addresses and other stuff. How does that fit what SIMH needs, which is an ethernet level send/receive service?

vmnet provides a virtual ethernet interface, SIMH is sending/receiving L2 frames over it. While the API documentation mentions hypervisors, SIMH needs the exact same thing as them. The L3 IP stuff vmnet does is for the purposes of subnet it uses for shared/host-only networking, where it handles the DHCP pool and NAT.

markpizz commented:
From my experience, this didn't work on all macOS versions that are in use by users.  Specifically, 10.15.7.

Additionally, the usage requires users to know something different than before.  When I raised this question, his answer was:

> Correct, like the other alternate ethernet backends, it's just prefixed (i.e. vmnet:host or vmnet:en0). It is a little different in that bridging and NAT are presented under the same interface (i.e. it can be used as a substitute for slirp or pcap).

That answer is wrong.  We don't have users saying slirp:xxxxx  They say nat:
Likewise, we don't have users saying: pcap:en0, they say eth0 or en0.

Additionally, the vmnet: NAT functionality is not useful for any of the most common use cases.  This is due to it depending on client systems getting IP addresses via DHCP, while almost no operating systems running on simh simulators have DHCP support built in to the operating systems.  It ALSO DOES NOT provide any way to actually create open inbound ports which are very commonly used by simh users.

pkoning2 commented:
I tried vmnet:en5 (which is a wired Ethernet dongle) and it worked fine.  Yes, pcap (the default) works also.  Personally I tend to choose pcap, but the existing SIMH has long offered a choice of several different mechanisms.  And yes, tun/tap on Mac is either close to or entirely dead.  My interpretation of the PR is that it offers a replacement for tun/tap that works, i.e., an alternate for people who for any reason do not want to use pcap.  I see nothing wrong with that.

On the NAT thing, I assume that the only mode that's actually useful is the bridged mode, because we're emulating an Ethernet NIC, not an IP device.

pmetzger commented:
Doesn't getting rid of this get rid of the easiest way for a modern Mac user to get networking support? It's very Mac specific. (I see no reason to care about the fact that it won't work on 10 or 15 year old versions of MacOS.)

pkoning2 commented:
It's been a while, but as far as I know installing pcap on a current Mac is a trivial task.

markpizz commented:
@pmetzger  said:
> Doesn't getting rid of this get rid of the easiest way for a modern Mac user to get networking support? It's very Mac specific. (I see no reason to care about the fact that it won't work on 10 or 15 year old versions of MacOS.)

I've got vmnet working on a mac that's from Late 2012 which is running the latest OS that hardware supports (10.15.7).

@pkoning2  said:
> It's been a while, but as far as I know installing pcap on a current Mac is a trivial task.

Trivial or otherwise, vmnet does a better job of providing network support for a simulator.  Specifically, when using a vmnet bridged connection, the host can talk directly to the simulated system and the simulated system can talk directly to the LAN and beyond.  Clearly the simulated system would need to use an IP address appropriate to the LAN.

My only suggestion is that the name of the device that vmnet uses should merely be the interface name since it functionally replaces pcap which never accepted pcap:en0 but always accepted the raw interface name.  Anyone with a configuration that worked previously will still have something that works.  With that in place, the reason most folks previously had to attempt tun connections really no longer exist.

pmetzger commented:
Apple has a policy at this point of removing mechanisms that they perceive to be security risks. Third party kernel device drivers are now dead. Tun and tap are dead. pcap is difficult to use unless the user knows enough to change the ownership of various files in /dev and I am unsure whether it will be provided in the future except as a way of listening to packets. The official way on a modern Mac to provide a virtual machine with access to the network is vmnet. There is no argument against allowing a Mac user to have access to that. It doesn't matter that other platforms can't use this mechanism; other platforms can't use the Windows support code but we still have Windows support code.

pkoning2 commented:
@NattyNarwhal sorry for the delay in sorting this out.

I consulted with the other steering group members.  The answer that came back is that I was too critical.  What we want to see is:
The default build (no magic parameters) works on all supported OS and versions, using the standard compiler etc. for that system.  
So what that means here is that things should build on non-Mac systems (no vmnet of course) and on Mac systems both those too old to support vmnet and the newer ones that do support it.  And on Mac that would be with the standard compiler clang.
Given that gcc is no longer standard on Mac OS, build failures with gcc are allowed (though of course avoiding them would be nice, but that is optional).  We do want vmnet support to work cleanly on old Mac OS that don't have vmnet.  All this applies to both build procedures: the makefile and cmake.

I would much prefer vmnet to be a choice for network access just like tap and others are, where you specify a prefix on the device to get that choice.  pcap for historic reasons doesn't have a prefix, but the others do.  So vmnet doesn't replace pcap, it is an additional choice.  The argument is that having to change all the init files when updating Open SIMH is extra work.  Sometimes it is unavoidable but in this case it need not be as far as I can tell.

If all this is agreeable, please submit a new PR with these items addressed.  And again, sorry for the detours.

pkoning2 commented:
BTW, on pcap and devices, I have not seen that.  I have a shiny new Mac OS and it supports pcap without any magic on my part.  I don't remember where it came from; might be the Wireshark kit, or maybe from "homebrew", but in either case, I just did the install and it worked.

quagly commented:
Does vmnet support decnet?

pmetzger commented:
For whomever cares: I have restored this functionality on my hardfork of SIMH, which I'm calling Zeerust. It should be up on Github in the next couple of days. I'm sorry, I needed this stuff and I couldn't wait for any longer.

pmetzger commented:
It looks like the functionality is actually a bit problematic because it requires running as root, so I've backed it out of my own fork temporarily. However, vmnet-helper or vmnet_broker would provide comparable functionality without that requirement. I'll be constructing an implementation that uses those.
